control-plane/connect-inject/controllers/endpoints/endpoints_controller.go

```diff
@@ -975,13 +975,41 @@ func (r *Controller) deregisterService(apiClient *api.Client, k8sSvcName, k8sSvc
 					errs = multierror.Append(errs, err)
 				}
 			}
+
+			if serviceDeregistered {
+				err = r.deregisterNode(apiClient, k8sSvcName, k8sSvcNamespace, nodeSvcs.Node.Node)
+				if err != nil {
+					r.Log.Error(err, "failed to deregister node", "svc", svc.Service)
+					errs = multierror.Append(errs, err)
+				}
+			}
 		}
 	}
 
 	return errs
 
 }
 
+// deregisterNode removes a node if it does not have any associated services attached to it.
+// When using Consul Enterprise, serviceInstancesForK8SServiceNameAndNamespace will search across all namespaces
+// (wildcard) to determine if there any other services associated with the Node. We also only search for nodes that have
+// the correct kubernetes metadata (managed-by-endpoints-controller and synthetic-node).
+func (r *Controller) deregisterNode(apiClient *api.Client, k8sServiceName, k8sServiceNamespace, nodeName string) error {
+	nodeServices, err := r.serviceInstancesForK8SServiceNameAndNamespace(apiClient, k8sServiceName, k8sServiceNamespace, nodeName)
```

Comment:
I wonder if this is the wrong function because it filters on the `k8sServiceName`, but to clean the node we need to know if ANY service is associated with the node.

Reply:
We are getting all the services associated with that node with:

If I am reading the filter correctly it is:

give me the node with:
 meta {
 "k8s-service" = "static-server"
 "k8s-namespace" = "default"
 "managed-by" = "endpoints-controller"
 }

`filter := fmt.Sprintf(`Meta[%q] == %q and Meta[%q] == %q and Meta[%q] == %q`,
		metaKeyKubeServiceName, k8sServiceName, constants.MetaKeyKubeNS, k8sServiceNamespace, metaKeyManagedBy, constants.ManagedByValue)`

`serviceList, _, err = apiClient.Catalog().NodeServiceList(nodeName, &api.QueryOptions{Filter: filter})`

I think we do it with a filter like this instead of node because Consul would be slow to return without the filter.